homeassistant/components/husqvarna_automower/coordinator.py

```diff
@@ -44,7 +48,7 @@ async def _async_update_data(self) -> dict[str, MowerAttributes]:
             self.ws_connected = True
         try:
             return await self.api.get_status()
-        except ApiException as err:
+        except (ApiException, AuthException) as err:
             raise UpdateFailed(err) from err
```

Comment:
Is it not needed to fix authentication, or is it a false positive

Reply:
The server was down on monday evening for some hours. I think, the affected client tried to renew the access token in that time (has to be done every 24h). So as the server was down, it was not possible to renew. Is it that, what you asked, for?

Reply:
He means if this shouldn't start a `reauth` flow instead of just reporting update failed.

Reply:
May be add a comment?

Reply:
I think you were right...
Today I had to do a reauth witch also throwed an `AuthException`:
```
024-04-16 18:15:25.344 ERROR (MainThread) [custom_components.husqvarna_automower.coordinator] Unexpected error fetching husqvarna_automower data: Access token failure: 400, message='Bad Request', url=URL('https://api.authentication.husqvarnagroup.dev/v1/oauth2/token')
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/aioautomower/auth.py", line 126, in _async_get_access_token
    return await self.async_get_access_token()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/husqvarna_automower/api.py", line 28, in async_get_access_token
    await self._oauth_session.async_ensure_token_valid()
  File "/usr/src/homeassistant/homeassistant/helpers/config_entry_oauth2_flow.py", line 523, in async_ensure_token_valid
    new_token = await self.implementation.async_refresh_token(self.token)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/config_entry_oauth2_flow.py", line 94, in async_refresh_token
    new_token = await self._async_refresh_token(token)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/config_entry_oauth2_flow.py", line 184, in _async_refresh_token
    new_token = await self._token_request(
                ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/config_entry_oauth2_flow.py", line 217, in _token_request
    resp.raise_for_status()
  File "/usr/local/lib/python3.12/site-packages/aiohttp/client_reqrep.py", line 1070, in raise_for_status
    raise ClientResponseError(
aiohttp.client_exceptions.ClientResponseError: 400, message='Bad Request', url=URL('https://api.authentication.husqvarnagroup.dev/v1/oauth2/token')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 315, in _async_refresh
    self.data = await self._async_update_data()
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/husqvarna_automower/coordinator.py", line 45, in _async_update_data
    return await self.api.get_status()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/aioautomower/session.py", line 145, in get_status
    mower_list = await self.auth.get_json(AutomowerEndpoint.mowers)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/aioautomower/auth.py", line 73, in get_json
    resp = await self.get(url, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/aioautomower/auth.py", line 66, in get
    resp = await self.request("get", url, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/aioautomower/auth.py", line 55, in request
    headers = await self.headers()
              ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/aioautomower/auth.py", line 132, in headers
    access_token = await self._async_get_access_token()
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/aioautomower/auth.py", line 128, in _async_get_access_token
    raise AuthException(f"Access token failure: {err}") from err
aioautomower.exceptions.AuthException: Access token failure: 400, message='Bad Request', url=URL('https://api.authentication.husqvarnagroup.dev/v1/oauth2/token')
```
So this PR would habe changed the behavior, that only `UpdateFailed` was risen, but  a reauth was required. I think I have to change the library here and don't raise an AuthException anymore:
https://github.com/Thomas55555/aioautomower/blob/086fe23d8edaf9a649db81f9e27fbba58718630c/src/aioautomower/auth.py#L155

The  `AuthException` should only be rissen here:
https://github.com/Thomas55555/aioautomower/blob/086fe23d8edaf9a649db81f9e27fbba58718630c/src/aioautomower/auth.py#L128

But I have to wait until #114500 is merged, until I can update the library.

Reply:
I had a look at this again. Now I don't think, that a change in the library is needed, and this PR is sufficient, because `AuthException` is always rissen: In the referenced issue, in my example above. And I also forced another token failure, which gave me this log:
```
2024-05-08 21:07:12.609 ERROR (MainThread) [homeassistant.components.husqvarna_automower.coordinator] Unexpected error fetching husqvarna_automower data
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/aioautomower/auth.py", line 126, in _async_get_access_token
    return await self.async_get_access_token()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/husqvarna_automower/api.py", line 28, in async_get_access_token
    await self._oauth_session.async_ensure_token_valid()
  File "/usr/src/homeassistant/homeassistant/helpers/config_entry_oauth2_flow.py", line 523, in async_ensure_token_valid
    new_token = await self.implementation.async_refresh_token(self.token)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/config_entry_oauth2_flow.py", line 94, in async_refresh_token
    new_token = await self._async_refresh_token(token)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/config_entry_oauth2_flow.py", line 184, in _async_refresh_token
    new_token = await self._token_request(
                ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/config_entry_oauth2_flow.py", line 217, in _token_request
    resp.raise_for_status()
  File "/usr/local/lib/python3.12/site-packages/aiohttp/client_reqrep.py", line 1070, in raise_for_status
    raise ClientResponseError(
aiohttp.client_exceptions.ClientResponseError: 400, message='Bad Request', url=URL('https://api.authentication.husqvarnagroup.dev/v1/oauth2/token')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 315, in _async_refresh
    self.data = await self._async_update_data()
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/husqvarna_automower/coordinator.py", line 46, in _async_update_data
    return await self.api.get_status()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/aioautomower/session.py", line 189, in get_status
    mower_list = await self.auth.get_json(AutomowerEndpoint.mowers)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/aioautomower/auth.py", line 73, in get_json
    resp = await self.get(url, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/aioautomower/auth.py", line 66, in get
    resp = await self.request("get", url, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/aioautomower/auth.py", line 55, in request
    headers = await self.headers()
              ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/aioautomower/auth.py", line 132, in headers
    access_token = await self._async_get_access_token()
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/aioautomower/auth.py", line 128, in _async_get_access_token
    raise AuthException(f"Access token failure: {err}") from err
aioautomower.exceptions.AuthException: Access token failure: 400, message='Bad Request', url=URL('https://api.authentication.husqvarnagroup.dev/v1/oauth2/token')
```

Reply:
You should implement a `reauth` flow to handle auth errors. Not just raise `UpdateFailed`

https://developers.home-assistant.io/docs/config_entries_config_flow_handler#reauthentication

Reply:
I think I was only looking at a subset of the commits. It seems right now